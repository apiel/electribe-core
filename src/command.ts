export const GET_CURRENT_PATTERN = [240, 66, 48, 0, 1, 35, 16, 247]; // F0,42,30,00,01,23,10,F7
