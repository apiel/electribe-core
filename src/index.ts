export * from './constant';
export * from './mod';
export * from './osc';
export * from './electribe';
export * from './hex';
export * from './codec';
