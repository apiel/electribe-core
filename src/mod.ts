export const MOD = [
    {
        name: 'EG+ Filter',
        source: 'AD Envelope (positive)',
        destination: 'Filter Cutoff',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ Pitch',
        source: 'AD Envelope (positive)',
        destination: 'Oscillator Pitch',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ OSC',
        source: 'AD Envelope (positive)',
        destination: 'Oscillator Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ Level',
        source: 'AD Envelope (positive)',
        destination: 'Amp Level',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ Pan',
        source: 'AD Envelope (positive)',
        destination: 'Pan',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ IFX',
        source: 'AD Envelope (positive)',
        destination: 'IFX Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG+ BPM Filter',
        source: 'AD Envelope (positive)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG+ BPM Pitch',
        source: 'AD Envelope (positive)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG+ BPM OSC',
        source: 'AD Envelope (positive)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG+ BPM Level',
        source: 'AD Envelope (positive)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG+ BPM Pan',
        source: 'AD Envelope (positive)',
        destination: 'Pan',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG+ BPM IFX',
        source: 'AD Envelope (positive)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- Filter',
        source: 'AD Envelope (negative)',
        destination: 'Filter Cutoff',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- Pitch',
        source: 'AD Envelope (negative)',
        destination: 'Oscillator Pitch',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- OSC',
        source: 'AD Envelope (negative)',
        destination: 'Oscillator Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- Level',
        source: 'AD Envelope (negative)',
        destination: 'Amp Level',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- Pan',
        source: 'AD Envelope (negative)',
        destination: 'Pan',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- IFX',
        source: 'AD Envelope (negative)',
        destination: 'IFX Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'EG- BPM Filter',
        source: 'AD Envelope (negative)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- BPM Pitch',
        source: 'AD Envelope (negative)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- BPM OSC',
        source: 'AD Envelope (negative)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- BPM Level',
        source: 'AD Envelope (negative)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- BPM Pan',
        source: 'AD Envelope (negative)',
        destination: 'Pan',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'EG- BPM IFX',
        source: 'AD Envelope (negative)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'LFOTri Filter',
        source: 'LFO (triangle)',
        destination: 'Filter Cutoff',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTri Pitch',
        source: 'LFO (triangle)',
        destination: 'Oscillator Pitch',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTri OSC',
        source: 'LFO (triangle)',
        destination: 'Oscillator Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTri Level',
        source: 'LFO (triangle)',
        destination: 'Amp Level',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTri Pan',
        source: 'LFO (triangle)',
        destination: 'Pan',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTri IFX',
        source: 'LFO (triangle)',
        destination: 'IFX Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'LFOTriB Filter',
        source: 'LFO (triangle)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'LFOTriB Pitch',
        source: 'LFO (triangle)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'LFOTriB OSC',
        source: 'LFO (triangle)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'LFOTriB Level',
        source: 'LFO (triangle)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'LFOTriB Pan',
        source: 'LFO (triangle)',
        destination: 'Pan',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'LFOTriB IFX',
        source: 'LFO (triangle)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB Filter',
        source: 'LFO (up-saw)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB Pitch',
        source: 'LFO (up-saw)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB OSC',
        source: 'LFO (up-saw)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB Level',
        source: 'LFO (up-saw)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB Pan',
        source: 'LFO (up-saw)',
        destination: 'Pan',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawUpB IFX',
        source: 'LFO (up-saw)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB Filter',
        source: 'LFO (down-saw)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB Pitch',
        source: 'LFO (down-saw)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB OSC',
        source: 'LFO (down-saw)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB Level',
        source: 'LFO (down-saw)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB Pan',
        source: 'LFO (down-saw)',
        destination: 'Pan',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SawDwnB IFX',
        source: 'LFO (down-saw)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB Filter',
        source: 'LFO (up-square)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB Pitch',
        source: 'LFO (up-square)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB OSC',
        source: 'LFO (up-square)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB Level',
        source: 'LFO (up-square)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB Pan',
        source: 'LFO (up-square)',
        destination: 'Pan',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquUpB IFX',
        source: 'LFO (up-square)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB Filter',
        source: 'LFO (down-square)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB Pitch',
        source: 'LFO (down-square)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB OSC',
        source: 'LFO (down-square)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB Level',
        source: 'LFO (down-square)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB Pan',
        source: 'LFO (down-square)',
        destination: 'Pan',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'SquDwnB IFX',
        source: 'LFO (down-square)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: true,
    },
    {
        name: 'S&HBPM Filter',
        source: 'LFO (sample & hold)',
        destination: 'Filter Cutoff',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'S&HBPM Pitch',
        source: 'LFO (sample & hold)',
        destination: 'Oscillator Pitch',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'S&HBPM OSC',
        source: 'LFO (sample & hold)',
        destination: 'Oscillator Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'S&HBPM Level',
        source: 'LFO (sample & hold)',
        destination: 'Amp Level',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'S&HBPM Pan',
        source: 'LFO (sample & hold)',
        destination: 'Pan',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'S&HBPM IFX',
        source: 'LFO (sample & hold)',
        destination: 'IFX Edit',
        bpmSync: true,
        keySync: false,
    },
    {
        name: 'Random Filter',
        source: 'LFO (random)',
        destination: 'Filter Cutoff',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'Random Pitch',
        source: 'LFO (random)',
        destination: 'Oscillator Pitch',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'Random OSC',
        source: 'LFO (random)',
        destination: 'Oscillator Edit',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'Random Level',
        source: 'LFO (random)',
        destination: 'Amp Level',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'Random Pan',
        source: 'LFO (random)',
        destination: 'Pan',
        bpmSync: false,
        keySync: false,
    },
    {
        name: 'Random IFX',
        source: 'LFO (random)',
        destination: 'IFX Edit',
        bpmSync: false,
        keySync: false,
    },
];
