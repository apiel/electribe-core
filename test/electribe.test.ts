import { readFileSync } from 'fs';
import { parseMessage } from '../src';

 // to fix after refactoring!

describe('electribe', () => {
    let data: number[] = [];
    beforeAll(() => {
        const output = readFileSync(`${__dirname}/214.json`);
        data = JSON.parse(output.toString());
    });

    it('should do something', () => {
        const res = parseMessage(data);
        expect(res?.pattern).toEqual({
            alternate13_14: true,
            alternate15_16: true,
            beat: '16',
            chainRepeat: 0,
            chainTo: 0,
            chordSet: 3,
            gateArp: 1,
            key: 'G',
            keyId: 7,
            length: 4,
            level: 127,
            mfx: 'Looper',
            mfxHold: false,
            mfxId: 7,
            name: '300Miles 2',
            scale: 'Dorian',
            scaleId: 2,
            swing: 10,
            tempo: 1020,
            part: [...Array(16)].map((_, i) => ({
                name: `part ${i + 1}`,
                oscId: 263,
                osc: { name: 'Ahaa..', type: 'Voice' },
                oscEdit: 95,
                pitch: 63,
                glide: 38,
                filterId: 9,
                filter: { name: 'P5 HPF', type: 'HPF' },
                cutoff: 70,
                resonance: 64,
                egInt: 24,
                modulationId: 61,
                modulation: {
                    name: 'S&HBPM Filter',
                    source: 'LFO (sample & hold)',
                    destination: 'Filter Cutoff',
                    bpmSync: true,
                    keySync: false,
                },
                modSpeed: 97,
                modDepth: 54,
                mfxSend: true,
                ampEG: true,
                level: 59,
                pan: 'R 40',
                attack: 45,
                decayRelease: 71,
                ifxOn: false,
                ifxId: 33,
                ifx: 'Phaser Manual',
                ifxEdit: 53,
            })),
        });
    });
});
