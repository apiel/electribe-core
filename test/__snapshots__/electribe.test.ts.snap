// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`electribe should do something 1`] = `
Object {
  "alternate13_14": true,
  "alternate15_16": true,
  "beat": "16",
  "chainRepeat": 0,
  "chainTo": 0,
  "chordSet": 3,
  "gateArp": 1,
  "key": "G",
  "keyId": 7,
  "length": 4,
  "level": 127,
  "mfx": "Looper",
  "mfxHold": false,
  "mfxId": 7,
  "name": "300Miles 2",
  "parts": Array [
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 1",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 2",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 3",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 4",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 5",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 6",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 7",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 8",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 9",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 10",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 11",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 12",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 13",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 14",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 15",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
    Object {
      "effect": Object {
        "edit": 53,
        "id": 33,
        "name": "Phaser Manual",
        "on": false,
      },
      "envelope": Object {
        "ampEG": true,
        "attack": 45,
        "decayRelease": 71,
        "level": 59,
        "pan": "R 40",
      },
      "filter": Object {
        "cutoff": 70,
        "egInt": 24,
        "id": 9,
        "name": Object {
          "name": "P5 HPF",
          "type": "HPF",
        },
        "resonance": 64,
      },
      "modulation": Object {
        "depth": 54,
        "id": 61,
        "name": Object {
          "bpmSync": true,
          "destination": "Filter Cutoff",
          "keySync": false,
          "name": "S&HBPM Filter",
          "source": "LFO (sample & hold)",
        },
        "speed": 97,
      },
      "name": "part 16",
      "oscillator": Object {
        "edit": 95,
        "glide": 38,
        "id": 263,
        "name": Object {
          "name": "Ahaa..",
          "type": "Voice",
        },
        "pitch": 63,
      },
      "settings": Object {
        "lastStep": 16,
        "mfxSend": true,
        "partPriority": "Normal",
        "voiceAssign": "Mono 2",
      },
    },
  ],
  "scale": "Dorian",
  "scaleId": 2,
  "swing": 10,
  "tempo": 1020,
}
`;
